web/packages/design/src/DataTable/Pager/ClientsidePager/useClientsidePager.ts

```diff
@@ -0,0 +1,61 @@
+/*
+Copyright 2023 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import { FetchStatus } from '../../types';
+
+export default function useClientsidePager({
+  data = [],
+  paginatedData = [],
+  currentPage,
+  pageSize,
+  ...props
+}: Props) {
+  const currentPageData = paginatedData[currentPage] || [];
+  const searchFrom = currentPage * pageSize;
+
+  const from = data.indexOf(currentPageData[0], searchFrom);
+  const to = data.lastIndexOf(
+    currentPageData[currentPageData.length - 1],
+    searchFrom + pageSize - 1
+  );
+
+  const count = data.length;
+
+  const isNextDisabled = to === data.length - 1;
+  const isPrevDisabled = currentPage === 0;
+
+  return {
+    from,
+    to,
+    count,
+    isNextDisabled,
+    isPrevDisabled,
+    ...props,
+  };
+}
+
+export type Props = {
+  nextPage: () => void;
+  prevPage: () => void;
+  data: any[];
+  paginatedData?: Array<Array<any>>;
+  currentPage?: number;
+  pageSize?: number;
+  onFetchMore?: () => void;
+  fetchStatus?: FetchStatus;
+};
+
+export type State = ReturnType<typeof useClientsidePager>;
```

Comment:
I'm not a fan of this. It's basically saying "the type is whatever it is" which kind of defeats the purpose of using the type system to be explicit about what data we're going to return. What do you think about manually typing this?

Reply:
I agree, however this is the same pattern we use in all our `Container` hooks. This pattern will be phased out as we move away from using the `Container` and hooks pattern as @ryanclark suggested we do. I'm not sure if it makes sense to break that pattern just for this hook in this PR, what do you think?

Reply:
I think we should explicitly type this one, and start to do that more and more going forward.